docs/testing-strategy.md

```diff
@@ -0,0 +1,154 @@
+- they should navigate the application using the same means as a user, for
+  example clicking links or submitting forms;
+- they shouldn't test dead ends in a user flow (such as validation or
+  permission issues); these are simpler tested in
+  [request tests](#request-tests);
+- incidental behaviour (for example flash messages) shouldn't be tested, this
+  can be tested at a higher granularity in [unit](#unit-tests) or
+  [request](#request-tests) tests;
+- they shouldn't test the effects a test has on the database, only user visible
+  signs of success should be asserted on;
+- they shouldn't mock application code;
+- they may mock and/or assert that particular API calls are made to external
+  services;
+- they should be written in the [readable feature test
+  style][readable-feature-tests] popularised by Futurelearn.
+
+## Request tests
+
+The purpose of these tests is to test the HTTP API for the application from
+a machine perspective. They are used to determine that for a given input and
+application state a particular response is returned. These differ from feature
+tests in that they test a single endpoint at a time and can exhaust the logical
+outcomes of that endpoint where some situations may be difficult for
+a user to fall into.
+
+Characteristics of request tests:
+
+- they should test the variety of responses an endpoint returns, checking
+  aspects of the response such as status code and any flash messages;
+- they shouldn't test responses where there isn't specific logic written for
+  the scenario;
+- they should assert against the HTTP responses involved in the request, and
+  not side-effects such as database changes, these are better suited to [unit
+  testing](#unit-tests) an [interactor][interactors];
+- they should focus on the effects of a single endpoint with a single HTTP
+  method at a time, for example `POST /documents`, for multiple endpoints
+  consider a [feature test](#feature-tests);
+- they act as integration tests and thus shouldn't mock application code;
+- they can mock external API calls, but these should not be asserted against -
+  this is better suited to [unit tests](#unit-tests);
+- they shouldn't be used to extensively test HTML output, this is better suited
+  by [view tests](#view-tests);
+- they may not be necessary for places where a response doesn't have logic
+  and already has coverage provided by a feature test.
+
+## View tests
+
+The purpose of view tests is to assert that the expected HTML is rendered in a
+particular scenario. Typically it isn't necessary to have view tests as the
+use of complex logic is discouraged in views and often aspects of views are
+implicitly tested in feature or request tests. However when it is suited for
+us to test particular HTML output they are the most appropriate choice.
+
+Characteristics of view tests:
+
+- they should make assertions based on logic in the view and not be used to
+  assert that HTML is as we wrote it;
```

Comment:
Could you elaborate on this line? Based on https://github.com/alphagov/content-publisher/pull/1592, I'd be tempted to add:

```suggestion
  assert that HTML is as we wrote it (for example, assert that a selector contains specific text,
  rather than testing for the presence of all classes, IDs and data attributes a node might have);
```